### PR TITLE
Skin layout revamp dev

### PR DESCRIPTION
**Pull request by ferranlala**
Summary of changes:
-Removed the layout option buttons of the top of LateNight
-Added layout options to view menu. Rearranged view menu.
-Added the option to hide the crossfader to LateNight and Shade
-Modified LateNight so that the library cannot be hidden when the skin is in medium or full size.

### CODE REVIEW

**Comment by ywwg** [comment]
I'll try to take a look at this later, thanks

**Comment by Be-ing** [comment]
Nice, this feels better. A few small issues I notice in LateNight are that the "W" at the end of "PREVIEW" in the preview deck is cut off and there are no tall waveforms anymore.
mixxx-latenight-pr639

**Comment by Be-ing** [comment]
With the skin option buttons removed from the top, the clock looks a bit lonely up there. Maybe the clock could be aligned to the left of that bar and the Mixxx logo centered?

File: res/skins/LateNight/mixer.xml [hunk 1]
```diff
@@ -147,6 +147,10 @@
                                 </Children>
                             </WidgetGroup>
                         </Children>
+                            <Connection>
+						<ConfigKey>[Master],show_xfader</ConfigKey>
+							<BindProperty>visible</BindProperty>
+						</Connection>
```
Comment (daschuer) [inline res/skins/LateNight/mixer.xml]: Nitpick: replace tabs with with four spaces.

File: res/skins/Shade/mixer_panel.xml [hunk 2]
```diff
@@ -137,6 +137,10 @@
 								<ConfigKey>[Master],crossfader</ConfigKey>
 								<EmitOnDownPress>false</EmitOnDownPress>
 							</Connection>
+                            <Connection>
+                                <ConfigKey>[Master],show_xfader</ConfigKey>
+                                <BindProperty>visible</BindProperty>
+                            </Connection>
```
Comment (daschuer) [inline res/skins/Shade/mixer_panel.xml]: Blame on me, for the bad alignment.

**Comment by daschuer** [comment]
What is the use-case for removing the xFader in Shade. Any idea how to use the empty area? 
This will allow a Numark NDX three button laylout or something fancy like vinyl break effects.

**Comment by ferranpujolcamins** [comment]
In the initial discussion, Owen replied:

"If you add "hide crossfader" to the menu you'll also have to add support
for hidden
crossfader to Deere and Shade as well. Otherwise people using those
skins will file bugs saying they tried to hide the crossfader and it's
not working."

So I've just pleased him :)

We could just make the faders and the meters longer.

2015-07-07 20:31 GMT+02:00 Daniel Schürmann <notifications@github.com>:

**Comment by daschuer** [comment]
LateNight: (This is all a matter of taste. @ywwg will have the final word) 
* The Clock bar is not aligned with the empty bar on top of the decks. 
* On my Cinnamon desktop the skin comes up with the layout buttons below my display. How about move the footer onto the top and merge it with the clock and the CPU widget? 
* Why did you remove the 4/2 Decks button? Is this more a config option instead of a live feature? 
* What is the use-case for hide the EQs? The saved screen seams to be unused. 
* The preview deck looks sonhow foreign. Can we integrate it to the Mixxer area? Big Mixer = Small preview deck and vice verse.

**Comment by daschuer** [comment]
@ferranpujolcamins: Thank you for working on this. IMHO with some tweaks it suits to be part of 1.12.
How about the additional strings? Will this be a problem since we are already past beta?

**Comment by ferranpujolcamins** [comment]
-I already had planned to reduce the minimum height of the skin (it will be in another PR). @daschuer could you please tell me which is your screen resolution? Also, could you find out which is the maximum "minimum height" of the skin that allows you to maximize Mixxx in Cinnamon?

-I personally think the 2/4 decks button is something you set up once and don't touch during a performance. If someone heavily uses it I don't mind to add the button to the bar.

-The show/hide EQ was already there before this PR.

-The preview deck can surely be improved, in general it is our duty now to improve the skins to take advantage of every bit of space a hidden feature leaves, but this is out of scope of this PR.

- @Be-ing what do you mean by "tall waveforms"? When the mixer is hidden the waveforms appear a bit taller, just like in trunk. So nothing has changed here. Also, I can't reproduce the text issue with the preview deck.

**Comment by ferranpujolcamins** [comment]
I've fixed the formatting issues.
I'm investigating what can be done with the clock and the latency meter. I'll send you screenshots.

**Comment by Be-ing** [comment]
> @Be-ing what do you mean by "tall waveforms"? When the mixer is hidden the waveforms appear a bit taller, just like in trunk. So nothing has changed here.

You're right. Sorry, I got confused because the 1.12 branch currently has the option to toggle big/small waveforms independent of showing the mixer (which I think is kinda awkward and unnecessary).

> Also, I can't reproduce the text issue with the preview deck.

I am using KDE Plasma 5.3 on Fedora 22.

**Comment by Be-ing** [comment]
Alternatively, the clock and CPU meter could be moved near the Mixxx logo. I suspect it would look better if they stay up top though.

**Comment by ywwg** [comment]
> currently has the option to toggle big/small

this was removed a while ago, you might have to update to see the difference.

**Comment by ferranpujolcamins** [comment]
Can't we get rid of the Mixxx logo?

**Comment by ferranpujolcamins** [comment]
What about this?
captura de pantalla de 2015-07-07 23 13 00

**Comment by Be-ing** [comment]
Hmm, it could be a little weird/confusing having the horizontal CPU meter adjacent to the master VU meter. What if the clock and CPU meter were below the Mixxx logo?

**Comment by daschuer** [comment]
Mixxx logo: See the discussion here:  
https://github.com/mixxxdj/mixxx/commit/0c6c6ffb1cfae7e993d71620eee54c4a82032470#commitcomment-10058233

Maximize is not possible on my 1366x786 cinnamon screen. 
I use one head panel with the size of 25 pixel. Header bar and tool bar are also 25 Pixel.
So I think the maximum skin height 736 for me.

**Comment by ferranpujolcamins** [comment]
I don't really like this way.
captura de pantalla de 2015-07-07 23 17 58

**Comment by Be-ing** [comment]
Yeah, that is pretty odd now that I can see it. What about putting the clock in the top right of that section and the CPU meter in the top left?

**Comment by ferranpujolcamins** [comment]
So @daschuer would LateNight maximize in a "stock" Cinnamon desktop?

**Comment by ferranpujolcamins** [comment]
I like this :)
captura de pantalla de 2015-07-07 23 34 20

Even though the master meters might benefit from a little of padding on their right. However there's another PR that modifies their placement, so we shall wait for it to be applied before we make more tweaks to this.

**Comment by Be-ing** [comment]
Yeah, that's better. I'm not sure if the horizontal divider between the clock and master VU meters is necessary.

**Comment by ferranpujolcamins** [comment]
I don't really care.
captura de pantalla de 2015-07-07 23 39 49

I'll wait to see what other people say and commit this.

**Comment by ywwg** [comment]
> Can't we get rid of the Mixxx logo?

absolutely not

**Comment by ywwg** [comment]
I think this is a really good start!  How did you get the view options to gray out on the Deere skin?

* Show Spinnies disappeared.  We have users who report much better performance on extremely crappy hardware if they can disable spinnies.  That should be one of the View options.  (note that the show spinny icon is still in each deck, but that's horrible and we need to change that toggle switch to something else)
* I really think 2/4 decks needs to be on the skin itself.  I find myself switching that setting a lot, even during a set.
* Same with preview deck
* Maybe put the clock and CPU meter in the bottom button bar?  Also maybe label the cpu meter with "CPU" (even though it's not technically CPU usage, it's more understandable than "load" or "latency"
* When turning off crossfader and eqs, it would be nice if the mixer was at the top of the window, not the bottom.

**Comment by Be-ing** [comment]
> Maybe put the clock and CPU meter in the bottom button bar? Also maybe label the cpu meter with "CPU" (even though it's not technically CPU usage, it's more understandable than "load" or "latency"

That seems like a better idea than the last screenshot.

**Comment by ferranpujolcamins** [comment]
Ok, I've moved the clock and the latency meter to the bottom bar and added the 2/4 decks and the show preview deck buttons. I've also added a "CPU" label to the latency meter. I've also unified the different bottom bars that existed because they were almost identical anyway.

I've changed the mixer alignment to top, in general it looks nice to me. However, this makes the background of the crossfader too small. There's already a PR (https://github.com/mixxxdj/mixxx/pull/634) that removes the crossfader border and background, which I think is the way to go. Shall I merge this branch?

I agree about the show spinnies button. I've thought we could change that with a crossfader assign switch, like the ones there are already in the samplers.

Deere and Shade can hide each spinny individually. This means there will be 4 "show spinny on deck N" view menu options. However, LateNight handles all the spinnies at the same time with a special CO. How do we handle this? Right now the view menu options can't change according to which skin is loaded, isn't it? A few days ago I attempted to make the spinnies in LateNight individually "hiddable". However this led to the waveforms playing position not being aligned between decks with the spinny visible and decks with the spinny hidden. What's the way to go here? 

By the way, there's one problem: when on med or full mode, the show library CO does nothing. It behaves like this for two reasons:
1-It didn't make sense to be able to hide the library if the space was to be unused anyway.
2-The show library CO is a trigger for a widget stack when LateNight is in small mode.
If the last time LateNight was in small mode it had the sampler page selected (or fx or mic page), on the next Mixxx run, the show library CO is set to 0 because the selected tab is another one. This happens even if LateNight is in med mode, because the fx_samplers_container_small template is hidden but is still there. What this caused is that if LateNight was in med or full mode on Mixxx start, the library was hidden even if it was visible on the last Mixxx run. Disconnecting the library from the show library CO in med and full mode makes this bug disappear.

**Comment by ferranpujolcamins** [comment]
PS: Maybe we could move the bar to the top of the skin. What do you think?

**Comment by daschuer** [comment]
> There's already a PR (#634) that removes the crossfader border and background, which I think is the way to go. Shall I merge this branch?

No, lets avoid criss cross merge. We also need the permission to merge it and a clean rebase. 
Is https://github.com/mixxxdj/mixxx/pull/634 polished for merge? I will ask @synthor.

**Comment by daschuer** [comment]
> PS: Maybe we could move the bar to the top of the skin. What do you think?

+1 this will solve issues on small screens. I like the clock centered on top

**Comment by ywwg** [comment]
> PS: Maybe we could move the bar to the top of the skin. What do you think?

I'd prefer to keep it on the bottom for now.  I'll take a look at the clock placement and see if it's working.



> However, LateNight handles all the spinnies at the same time with a special CO. How do we handle this?

personally I don't see a need for spinny toggling on individual decks.  It doesn't really make sense as a per-deck selection.  So I'd prefer to adapt all of the skins to act like LateNight in that regard.

**Comment by ferranpujolcamins** [comment]
Ok, this should be all.

Current issues:

-I couldn't correctly map the keyboard shortcuts for show/hide cover art. I don't know how to map the key next to the - (US layout). Right know I could only make it work with Spanish, Italian and Russian.

-PR 634 needs to be merged for the crossfader to look good in LateNight.

-We still have the "sometime, the show library button on LateNight does nothing" issue.

-I've deliberately not updated Deere because it needs major rework, so we will apply all this improvements to it when we put some love to it in the near future.

-Shade needs some love. Deere needs a whole lot of love. I'll probably put some time on them now that I'm finally on holidays. I thought I could make new cool graphics for Deere, something similar (but not equal) to this:
http://medialoot.com/blog/how-to-create-a-detailed-audio-rotary-knob-control-in-photoshop-illustrator/
Does the knobs really need to be vector graphics? If they are bitmaps it is much easier to make them look fantastic, like on the link.

-Something else?

**Comment by synthor** [comment]
Wasn't able to repair my branch. So I created a new one of 1.12 and cherry-picked my four commits on it. See #641

**Comment by Be-ing** [comment]
This is looking better and better. :) I like the replacement of the spinny toggle buttons with crossfader assignment buttons. This is a better use of the space now that Mixxx supports 4 decks.

I am in favor of moving the bar to the top. I also have issues with the buttons being cut off on my laptop screen. Before, I thought that was a bug in KWin 5, not Mixxx.
I think it would help to put a little spacer between the "LIBRARY" and "MIC/AUX" buttons so the "MIC/AUX", "EFFECTS", and "SAMPLERS" buttons appear as a group.

>I couldn't correctly map the keyboard shortcuts for show/hide cover art. I don't know how to map the key next to the - (US layout). Right know I could only make it work with Spanish, Italian and Russian

Ctrl and = is working for me on my en_US keyboard.

> We still have the "sometime, the show library button on LateNight does nothing" issue.

There also still https://bugs.launchpad.net/mixxx/+bug/1462061 which is really annoying.

**Comment by ferranpujolcamins** [comment]
I really had no idea what I was doing when I set the keyboards shortcuts, so we really need someone to confirm they work as expected.

> I also have issues with the buttons being cut off on my laptop screen.
I'm working towards the reduction of the minimum height of LateNight. A first step is building a slider that can be resized, so I could borrow a few more pixels from the pitch slider. I've asked for help about this in the mailing list so if you know something about how the skinning engine behaves....:)

@ywwg , of all the issues we've mentioned in the discussion so far, which ones are blocking the merge of the PR? I want to start to work on other non-related skin improvements.

**Comment by synthor** [comment]
captura de pantalla de 2015-07-07 23 39 49

> I'll wait to see what other people say and commit this.

In 4ch mode it would be a problem with the master meter. I placed it centered below the master knobs. Exactly in row with the channel meters.

**Comment by daschuer** [comment]
> -I've deliberately not updated Deere because it needs major rework, so we will apply all this improvements to it when we put some love to it in the near future.

@esbrandt (the maintainer of Deere).  Can you give a status update? 

> -Shade needs some love. 

Yes. Do you have a special plan for it? 

> Deere needs a whole lot of love. I'll probably put some time on them now that I'm finally on holidays. I thought I could make new cool graphics for Deere, something similar (but not equal) to this:
http://medialoot.com/blog/how-to-create-a-detailed-audio-rotary-knob-control-in-photoshop-illustrator/

@esbrandt: (please correct me if I am wrong, I'm no design expert.) 
IMHO the original Deere idea was to use a flat-design for it. Recently we have introduced not that flat siders, which breaks the pure flat-design for me. 
In any case, I like the idea to have a flat-design pre-installed.

Your button looks more like skeuomorphism. In the German wikipedia are two nice pictures comparing the two designs: https://de.wikipedia.org/wiki/Flat_Design
I like the idea to offer a skeuomorphism design as well. So if you can effort the time, feel to fork any skin and add your own flavour to it. 

The 1.11 Deere design is for me also a candidate to lift to 1.12. 

> Does the knobs really need to be vector graphics? If they are bitmaps it is much easier to make them look fantastic, like on the link.

Mixxx 1.12 supports svg and png buttons. 

> -Something else?

Do you have an idea to solve the pending effect issue? We have requests to allow chaining filter end echo effect using LateNight.

**Comment by synthor** [comment]
Yeah, the next thing I like to work on is the staticness of the effectchains in LateNight (and other skins also). Chains are pretty useless, when the artist can not configure them like he want and is limited in his creativity.

**Comment by daschuer** [comment]
Four decks + Mixer + EQs + Preview Decks: 
If you enable the xFader, there is a ugly black area below the xfader. 
Without the xfader and the preview deck, the Mixer ends aligned with the decks. It would be nice if we could keep the alignement, and just add the xFader below.

**Comment by synthor** [comment]
My commits would look like this:

lnsr

**Comment by ferranpujolcamins** [comment]
> If you enable the xFader, there is a ugly black area below the xfader.

Yeah synthor's commits fixes this.

> Without the xfader and the preview deck, the Mixer ends aligned with the decks. It would be nice if we could keep the alignement, and just add the xFader below.

I take note of this, but I'll fix that in another PR, this PR is not a major LateNight rework.

**Comment by synthor** [comment]
> Yeah synthor's commits fixes this.

:)

**Comment by ywwg** [comment]
This looks really good!  Make sure to pull in 1.12 so you get the crossfader update.

The only bug I see is that In the larger sizes, the library button is clickable but doesn't do anything.

I think the clock is fine where it is, but it might also work in the right corner along with CPU.  Up to you.

**Comment by ywwg** [comment]
> staticness of the effectchains in LateNight (and other skins also).

One option is to make a generic FX chain template that can be applied to all skins with minor color theming.

**Comment by ywwg** [comment]
> I want to start to work on other non-related skin improvements.

I think this set of improvements is fine for this PR.  we don't have to solve all the problems now

**Comment by Be-ing** [comment]
> One option is to make a generic FX chain template that can be applied to all skins with minor color theming.

This seems like it could  be a good idea to unify the effects UI. The variability between skins now is  awfully confusing and unintuitive.

**Comment by ferranpujolcamins** [comment]
> The only bug I see is that In the larger sizes, the library button is clickable but doesn't do anything.

It's a workaround for another bug (worse in my opinion). As I commented before:

**Comment by ferranpujolcamins** [comment]
Oh, I've just thought I could change the color or the "L" and "R" of the crossfader assign button to red and blue, it's a quick fix. The user could see at a glance where is a deck assigned. Would you like that?

**Comment by ferranpujolcamins** [comment]
Another possible quick fix: can I get rid of the striped backgrounds on the decks and samplers? I don't really like them .

**Comment by ywwg** [comment]
> The user could see at a glance where is a deck assigned. Would you like that?

leave that alone for now, we use red and blue for deck 1/2 vs 3/4 so it would be confusing to also use red/blue for right/left.  Or just make both the R and L yellow text -- there's no real need to color code.

Please put it back so the library button is not clickable -- it makes more sense than to have the button be broken.  Or figure out a way to skin it so it is always on but has a "disabled" appearance.  Like, flat with grayer text.

**Comment by ferranpujolcamins** [comment]
> Please put it back so the library button is not clickable -- it makes more sense than to have the button be broken. Or figure out a way to skin it so it is always on but has a "disabled" (unclickable) appearance. Like, flat with grayer text.

I can do that we will still have the same problem with the view menu check option. I think this can't completely be solved until we have a way to communicate to Mixxx which CO a skin offers.

**Comment by ywwg** [comment]
how is it that in Deere, the menu items are grayed out?

> can I get rid of the striped backgrounds on the decks and samplers

no thanks :).  Maybe in the future I'll change my mind but let's leave them for now.

**Comment by ferranpujolcamins** [comment]
> how is it that in Deere, the menu items are grayed out?

I see all of them not grayed out. I'm investigating why the CO not supported are still active.

**Comment by Be-ing** [comment]
How is progress on this?

**Comment by ferranpujolcamins** [comment]
Hi, I thing the only thing that still needs to be done is this:


> Please put it back so the library button is not clickable -- it makes more sense than to have the button be broken. Or figure out a way to skin it so it is always on but has a "disabled" (unclickable) appearance. Like, flat with grayer text.

I'll try to fix this tomorrow.

**Comment by ferranpujolcamins** [comment]
Done. What else is blocking this?

**Comment by Be-ing** [comment]
Could you make the crossfader assignment CO persistent? As an example use case, I may want to keep decks 1 & 2 assigned to left and right, but not assign 3 & 4 to the crossfader.

**Comment by Be-ing** [comment]
I am opposed to merging this unless https://bugs.launchpad.net/mixxx/+bug/1462061 is fixed. Without the ability to hide the library, that bug can't be worked around.

**Comment by Be-ing** [comment]
This is still a bit too tall for my screen 1366x768 laptop screen. :/
late-night-pr-screenshot

**Comment by ferranpujolcamins** [comment]
Hi @Be-ing , Shrinking the height of LateNight is yet another thing I'm planning to do.
I'm investigating the bug you've mentioned as well.

**Comment by daschuer** [comment]
What is current state of this? I am afraid the git branch is messed up since there are unrelated commits listed, and it is not in a merge-able state. Is it still a 1.12 candidate?

**Comment by ywwg** [comment]
> Is it still a 1.12 candidate?

No, let's keep this master-only.  I think LateNight is in fine shape for 1.12.

**Comment by ywwg** [comment]
(to be specific, the code changes and changes to other skins required by this alteration, coupled with the "nice-to-have" nature of the changes tip it in my mind to something for master.  Although the height issue in latenight is annoying, it's not really a blocker)

**Comment by ferranpujolcamins** [comment]
In fact this doesn't solve the issue of LateNight being too tall anyway.
It's true that this was a first step towards a solution (hide/show
crossfader), but it has mutated to a layout options re-factoring.

We still need to make a lot of work in order to offer the user a *complete*
("all" functions available in each skin), *consistent* (skins can't have a
very different work-flow for common tasks such as beat-looping, fx, or the
layout options discussed in this PR), *polished* (that endless list of
little things that can look and work better) and *beautiful* experience
with skins. Don't worry we will eventually make it for 1.13 or whatever
release.

This is going to be long, so let's just take a compromise on each of the
skin aspects above for now. Let's just try to give birth to 1.12 ASAP :)

2015-08-03 14:46 GMT+02:00 Owen Williams <notifications@github.com>:

**Comment by Be-ing** [comment]
@ferranpujolcamins, are you planning on finishing this PR for 2.1?

**Comment by ferranpujolcamins** [comment]
Sure, I have some free days. I'll take a look.
El dia 19/01/2016 4:58 a. m., "Be" <notifications@github.com> va escriure:

**Comment by ferranpujolcamins** [comment]
Well, actually not. Sorry.

2016-01-19 8:22 GMT+01:00 Ferran Pujol Camins <ferranpujolcamins@gmail.com>:

**Comment by ferranpujolcamins** [comment]
Superseded by https://github.com/mixxxdj/mixxx/pull/913, which merges to master cleanly.

**Pull request closed without merging**